contracts/fuses/ebisu/EbisuZapperCreateFuse.sol

```diff
@@ -0,0 +1,249 @@
+import {IActivePool} from "./ext/IActivePool.sol";
+import {IAddressesRegistry} from "./ext/IAddressesRegistry.sol";
+import {IPriceFeed} from "./ext/IPriceFeed.sol";
+import {LiquityMath} from "./ext/LiquityMath.sol";
+import {FuseStorageLib} from "../../libraries/FuseStorageLib.sol";
+import {EbisuMathLib} from "./lib/EbisuMathLib.sol";
+import {PlasmaVaultConfigLib} from "../../libraries/PlasmaVaultConfigLib.sol";
+import {IWethEthAdapter} from "./ext/IWethEthAdapter.sol";
+import {WethEthAdapterStorageLib} from "./lib/WethEthAdapterStorageLib.sol";
+import {WethEthAdapter} from "./ext/WethEthAdapter.sol";
+import {EbisuZapperSubstrateLib, EbisuZapperSubstrate, EbisuZapperSubstrateType} from "./lib/EbisuZapperSubstrateLib.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+struct EbisuZapperCreateFuseEnterData {
+    address zapper;
+    address registry;
+    uint256 collAmount;
+    uint256 ebusdAmount;
+    uint256 upperHint;
+    uint256 lowerHint;
+    uint256 flashLoanAmount;
+    uint256 annualInterestRate;
+    uint256 maxUpfrontFee;
+}
+
+struct EbisuZapperCreateFuseExitData {
+    address zapper;
+    uint256 flashLoanAmount;
+    uint256 minExpectedCollateral;
+    bool exitFromCollateral;
+}
+
+contract EbisuZapperCreateFuse is IFuseCommon {
+    using SafeERC20 for IERC20;
+    uint256 public immutable MARKET_ID;
+    address public immutable WETH;
+
+    // from zapper
+    uint256 public constant ETH_GAS_COMPENSATION = 0.0375 ether;
+
+    uint256 private constant MIN_ANNUAL_INTEREST_RATE = 1e16 / 2; // 0.5%
+    uint256 private constant MAX_ANNUAL_INTEREST_RATE = 250 * 1e16; // 250%
+    uint256 private constant UPFRONT_INTEREST_PERIOD = 7 days;
+    uint256 private constant ONE_YEAR = 365 days;
+    uint256 private constant DECIMAL_PRECISION = 1e18;
+    uint256 private constant MIN_DEBT = 2000e18;
+
+    error DebtBelowMin(uint256 debt);
+    error ICRBelowMCR(uint256 icr, uint256 mcr);
+    error NewOracleFailureDetected();
+    error UnknownExitType();
+    error UpfrontFeeTooHigh(uint256 fee);
+    error UnsupportedSubstrate();
+    error TroveAlreadyOpen();
+    error TroveNotOpen();
+    error WethEthAdapterNotFound();
+    error WethAddressNotValid();
+
+    event EbisuZapperCreateFuseEnter(
+        address zapper,
+        uint256 collAmount,
+        uint256 flashLoanAmount,
+        uint256 ebusdAmount,
+        uint256 troveId
+    );
+    event EbisuZapperCreateFuseExit(address zapper, uint256 ownerIndex);
+    event WethEthAdapterCreated(address adapterAddress, address plasmaVault, address weth);
+
+    constructor(uint256 marketId_, address weth_) {
+        if(weth_ == address(0)) revert WethAddressNotValid();
+        MARKET_ID = marketId_;
+        WETH = weth_;
+    }
+
+    function enter(EbisuZapperCreateFuseEnterData calldata data_) external {
+        // Storage cache
+        FuseStorageLib.EbisuTroveIds storage troveData = FuseStorageLib.getEbisuTroveIds();
```

Comment:
Move line 84 and 87 to line 113 - so troveData you need first time at 114 line - it is more gas efficient